### ISSUE TITLE
[Bug]: opentelemetry-sdk leaves `telemetry.sdk.*` resources empty

### ISSUE THREAD

**Issue by Ilmarii** [open-telemetry/opentelemetry-rust#1940]
### What happened?

The logs of the [example]( https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-appender-tracing/examples/basic.rs) do not contain `telemetry.sdk.*` resources, which are mark as required https://opentelemetry.io/docs/specs/semconv/resource/#telemetry-sdk

### API Version

0.24

### SDK Version

0.24

### What Exporter(s) are you seeing the problem on?

N/A

### Relevant log output

```shell
{
  "resourceLogs": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "log-appender-tracing-example"
            }
          }
        ]
      },
      "scopeLogs": [
        {
          "scope": {
            "name": "opentelemetry-appender-tracing",
            "version": "0.5.0"
          },
          "logRecords": [
            {
              "timeUnixNano": null,
              "time": null,
              "observedTimeUnixNano": 1721193214108507100,
              "observedTime": "2024-07-17 05:13:34.108",
              "severityNumber": 17,
              "severityText": "ERROR",
              "attributes": [
                {
                  "key": "event_id",
                  "value": {
                    "intValue": 20
                  }
                },
                {
                  "key": "user_name",
                  "value": {
                    "stringValue": "otel"
                  }
                },
                {
                  "key": "user_email",
                  "value": {
                    "stringValue": "otel@opentelemetry.io"
                  }
                },
                {
                  "key": "name",
                  "value": {
                    "stringValue": "my-event-name"
                  }
                }
              ],
              "droppedAttributesCount": 0
            }
          ]
        }
      ]
    }
  ]
}
```

**Comment by lalitb** [comment]
@Ilmarii 
The current behavior is as follows: If resources are specified during the creation of the `LoggerProvider`, those resources will be used as the final set and will not be merged with the default resources provided by the SDK. If no resources are specified during the `LoggerProvider` creation, the default set of resources, including `telemetry.sdk.*`, will be used. And since example code is providing the resources, they remain the final set. This approach appears to be consistent with the [specifications](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#resource-creation. Let me know if I am missing something?

**Comment by Ilmarii** [comment]
@lalitb Yes, you are totally correct. However, I think this API behavior is very unintuitive: in most cases the user wants to add resource attributes, not remove the default attributes.

**Comment by cijothomas** [comment]
>  I think this API behavior is very unintuitive: in most cases the user wants to add resource attributes, not remove the default attributes

yes, agree. Will look at addressing that. Thanks for sharing your feedback.

**Comment by lalitb** [comment]
This was discussed in community meeting. To summarize the discussion:
1. It is possible for user to merge the custom resource attributes to the SDK provided and pass it to `LoggerProvider`.
```rust
    let logger_provider = LoggerProvider::builder()
        .with_resource(Resource::default().merge(&Resource::new([KeyValue::new(
            "my-attr-key",
            "my-attr-value",
        )])))
        .with_simple_exporter(exporter)
        .build();
```
2. Providing additive behavior will loose the flexibility provided with the current approach, say if user wants to remove some of the resource attributes added by SDK. We can come up with different APIs e.g. `LoggerProviderBuilder::with_resource_exclusive()` and `LoggerProviderBuilder::with_resource_additive()` to support both behavior but is it worth adding the complexity.

3. The current approach also seem to be aligned to the [specs ](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#sdk-provided-resource-attributes) unless I am interpreting it incorrect :)

--clip
The SDK MUST provide access to a Resource with at least the attributes listed at [Semantic Attributes with SDK-provided Default Value](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#semantic-attributes-with-sdk-provided-default-value). This resource MUST be associated with a TracerProvider or MeterProvider if another resource was not explicitly specified.

If a user wants to combine custom attributes with the default resource, they can use [Merge](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge) with their custom resource or specify their attributes by implementing [Custom resource detectors](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment) instead of explicitly associating a resource.

**Comment by Ilmarii** [comment]
Thanks for the deep dive into the problem!
Perhaps even without changing the API we can improve the user experience with more explainable method docs.
I suggest something like this for `with_resource()` docs comment:

The `Resource` to be associated with this Provider. Will replace the previously added `Resource`, including default resources.
When calling this function, you are responsible for providing the attributes required by the OTel semantic convention. Hint: You can use `Resources::default()` for default values.

I also think it would be good to include the code example you provided.

**Comment by lalitb** [comment]
I agree, this can be documented better, with example. Perhaps we can use this issue to track improving it.

**Comment by scottgerring** [comment]
Hey @lalitb , this looks straightforward enough - documentation and example updates. If it's still relevant feel free to assign it to me!
